src/core/qgsvectorfilewriter.cpp

```diff
@@ -2039,6 +2045,10 @@ gdal::ogr_feature_unique_ptr QgsVectorFileWriter::createFeature( const QgsFeatur
       case QVariant::ULongLong:
         OGR_F_SetFieldInteger64( poFeature.get(), ogrField, attrValue.toLongLong() );
         break;
+      case QVariant::Bool:
+        OGR_F_SetFieldInteger( poFeature.get(), ogrField, attrValue.toInt() );
+        OGR_Fld_SetSubType( OGR_F_GetFieldDefnRef( poFeature.get(), ogrField ), OFSTBoolean );
```

Comment:
OGR_Fld_SetSubType() should be called before OGR_L_CreateField(). In your below test, the returned field should be QVariant.Bool instead of QVariant.Int

Reply:
Thank you for your input @rouault, I'm discovering the OGR API.

> OGR_Fld_SetSubType() should be called before OGR_L_CreateField()

I updated the code accordingly, let me know if it's ok for you.

>  In your below test, the returned field should be QVariant.Bool instead of QVariant.Int

Mmmm, even with this change, the field is still a QVariant.Int. I'm going to do some more digging.

Reply:
> Mmmm, even with this change, the field is still a QVariant.Int. I'm going to do some more digging.

Indeed, I see the OGR provider doesn't handle OFSTBoolean yet

Reply:
> Indeed, I see the OGR provider doesn't handle OFSTBoolean yet

I updated the OGR provider to manage the boolean subtype. This way, I'm finally have a QVariant::Bool type in my unit test.

Let me know if you have some comments or advices.